aptos-move/e2e-move-tests/src/tests/function_value_capture_option.rs

```diff
@@ -0,0 +1,367 @@
+// Copyright © Aptos Foundation
+// SPDX-License-Identifier: Apache-2.0
+
+//! Tests for capturing option type in function values.
+
+use crate::{assert_success, tests::common, MoveHarness};
+use aptos_framework::BuildOptions;
+use aptos_package_builder::PackageBuilder;
+use aptos_transaction_simulation::Account;
+use aptos_types::{
+    on_chain_config::FeatureFlag,
+    transaction::{ExecutionStatus, TransactionStatus},
+};
+use claims::assert_matches;
+use move_core_types::account_address::AccountAddress;
+
+#[test]
+fn test_vm_value_fv_capture_option_1() {
+    let mut h = MoveHarness::new();
+    let acc = h.new_account_at(AccountAddress::from_hex_literal("0x99").unwrap());
+
+    // test directly capture option type
+    h.enable_features(vec![], vec![FeatureFlag::ENABLE_CAPTURE_OPTION]);
+    let status = publish(
+        &mut h,
+        &acc,
+        r#"
+        module 0x99::m {
+            use std::option;
+
+            struct FunctionStore has key {
+                f: ||option::Option<u64> has copy+drop+store,
+            }
+
+            public fun id(x: option::Option<u64>): option::Option<u64> {
+                x
+            }
+
+            fun init_module(account: &signer) {
+                let v = option::none();
+                let f: ||option::Option<u64> has copy+drop+store = || id(v);
+                move_to(account, FunctionStore { f });
+            }
+        }
+        "#,
+    );
+    assert_matches!(
+        status,
+        TransactionStatus::Keep(ExecutionStatus::ExecutionFailure { .. })
+    );
+
+    h.enable_features(vec![FeatureFlag::ENABLE_CAPTURE_OPTION], vec![]);
+    let status = publish(
+        &mut h,
+        &acc,
+        r#"
+        module 0x99::m {
+            use std::option;
+
+            struct FunctionStore has key {
```

Comment:
what about this test case:

```
struct FunctionStore has key {
    f: |option::Option<u64>| has copy+drop+store,
}
```
because you check for arg types inside a function type, I believe this is not allowed in the current impl?

Reply:
yeah, as you mentioned above, since parameters do not appear in the layout, we can allow this?

Reply:
Yes, but I think the current impl does not?

Reply:
In any case we should probably have such a test case 👍